Use Crate's Postgres support, introduce config file

juliusv opened the pull request:
Fixes https://github.com/crate/crate_adapter/issues/14
Fixes https://github.com/crate/crate_adapter/issues/13
Fixes https://github.com/crate/crate_adapter/issues/12

juliusv commented on README.rst:
```diff
@@ -6,7 +6,7 @@ This is an adapter that accepts Prometheus remote read/write requests,
 for Prometheus.
 
-Requires CrateDB 2.2.0 or greater.
+Requires CrateDB 3.1.0 or greater.
```
Btw.,  CrateDB 3.1.0 is not out yet, but my nightly build that has the required features says `3.1.0-SNAPSHOT` on startup, so I assume it'll be 3.1.0 once it's released.

chaudum replied:
Yes, the BEGIN/COMMIT feature + the fix for the TIMESTAMPTZ type will be available in CrateDB 3.1.0.

chaudum commented:
Is it really best practice to add dependencies in Github? I'm not a Golang expert, but that does not feel right. Dependencies can be resolved when the project is built, right?

chaudum commented on crate.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+func (c crateEndpoint) write(ctx context.Context, r *crateWriteRequest) error {
+	_, err := c.pool.PrepareEx(ctx, "write_statement", `INSERT INTO metrics ("labels", "labels_hash", "timestamp", "value", "valueRaw") VALUES ($1, $2, $3, $4, $5)`, nil)
```
I think we should make the prepared statement name unique, e.g. suffix with timestamp.

chaudum replied:
Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

juliusv replied:
> I think we should make the prepared statement name unique, e.g. suffix with timestamp.

Reading https://godoc.org/github.com/jackc/pgx#ConnPool.PrepareEx and skimming the underlying implementation, it at least seems like it should be ok to always use the same name for the same statement string:

"PrepareEx is idempotent; i.e. it is safe to call PrepareEx multiple times with the same name and sql arguments. This allows a code path to PrepareEx and Query/Exec/Prepare without concern for if the statement has already been prepared."

> Question: In Go, how performant is it to allocate strings?
In other languages I would use a constant for the statement, as this part of the code is executed very often.

I'd be surprised if you could measure any difference (haven't tried yet), especially since this only happens once per remote-write-request, not per sample. But I think it's a good idea to put it into a `const` anyway.

chaudum replied:
Ok, cool. Sounds good to me.

juliusv commented:
@chaudum Vendoring (making a copy) of all dependencies is pretty standard in Go for non-experimental projects (in repos that represent final binaries, not in library repos). It's the only way to ensure that your code can still be built - and built reasonably reproducibly - in X amount of time.

chaudum commented on crate.go:
```diff
@@ -0,0 +1,147 @@
+		Port:     ep.Port,
+		User:     ep.User,
+		Password: ep.Password,
+		Database: ep.Database,
```
`Database: ep.Schema,`

CrateDB does not know the concept of a database in terms of other RDBMSs.
We should always say `schema`, if the equivalent of database in Postgres is meant.

juliusv replied:
Good to know, will call it schema on our end (can't change the `pgx.ConnConfig` struct of course).

juliusv commented:
Some more background on this:

https://peter.bourgon.org/go-best-practices-2016/#dependency-management
https://peter.bourgon.org/go-in-production/#dependency-management

chaudum commented on crate.go:
```diff
@@ -0,0 +1,147 @@
+			[]interface{}{
+				a.labels,
+				a.labelsHash,
+				// TODO: Find non-string way of encoding timestamps.
```
We will add a QA test for all types https://github.com/crate/crate-qa/blob/master/tests/client_tests/go/bulk_operations.go

juliusv commented:
(note that the details around the mechanics of vendoring in the second link are pretty outdated)

chaudum commented on server.go:
```diff
@@ -290,37 +266,28 @@ func (ca *crateAdapter) handleRead(w http.ResponseWriter, r *http.Request) {
+				labelsHash: fp,
+				timestamp:  time.Unix(0, s.Timestamp*1e6).UTC(),
+				value:      s.Value,
+				// Crate.io can't handle full NaN values as required by Prometheus 2.0,
```
We would like to fix this. What exactly does CrateDB not support? What does "full NaN" mean?

juliusv replied:
This is a comment that already existed in `master`, but I'm sure it's about the difference between "normal" NaN values and signalling NaN values, see https://github.com/prometheus/prometheus/blob/master/pkg/value/value.go and https://en.wikipedia.org/wiki/NaN#Signaling_NaN.

Prometheus uses one type of signalling NaN to explicitly mark stale samples. For this to work, the underlying database has to support storing the exact bit pattern that is sent to it. Not sure if CrateDB can do that via Postgres now?

chaudum commented on server.go:
```diff
@@ -365,50 +332,64 @@ func (ca *crateAdapter) handleWrite(w http.ResponseWriter, r *http.Request) {
+	Port             uint16 `yaml:"port"`
+	User             string `yaml:"user"`
+	Password         string `yaml:"password"`
+	Database         string `yaml:"database"`
```
Please name the setting "schema". See first comment

juliusv commented on crate.go:
```diff
@@ -0,0 +1,147 @@
+				a.labels,
+				a.labelsHash,
+				// TODO: Find non-string way of encoding timestamps.
+				a.timestamp.Format("2006-01-02 15:04:05.000-07"),
```
Great, will test and update here if it works.

juliusv replied:
Hmm, I tried two variants here, both based against the latest night build based on 33f0c20, which includes commit 90e4343.

This one gives the error: `cannot convert {2018-07-03 13:54:21.99 +0000 UTC 2 none} to Timestamptz`:

```golang
pgtype.Timestamptz{Time: a.timestamp, Status: pgtype.Present}
```

This pointer variant seems to succeed, but all timestamps in the DB end up being `946684800000 (2000-01-01T00:00:00.000Z)`:

```golang
&pgtype.Timestamptz{Time: a.timestamp, Status: pgtype.Present}
```

Do you perhaps have a working example in Go?

juliusv replied:
If I just try `a.timestamp` btw., which is just a `time.Time`, I also get only `2000-01-01T00:00:00.000` inserted.

chaudum replied:
Think we'll need to investigate further. We have a test for inserting only, but we don't check the actual value that we store. We should've done that really.
Think we can keep the "workaround" (as it's universally working across versions) until we've a fix for the correct timestamp insert.

juliusv replied:
Ok cool, I pushed the other fixups now.

chaudum commented:
@juliusv Regarding vendoring. Happy to learn something new :) However, makes perfectly sense, as there aren't any mirrored Go package repositories that would guarantee long-term availability.

chaudum commented:
@joemoe this PR looks good to me. should we discuss release strategy before we merge into master?

chaudum commented:
@juliusv For the future we may want to look at https://github.com/golang/dep for dependency management/vendoring. Do you know this project?

chaudum commented:
@juliusv Could you squash the last 3 commits + the review fixups into a single commit, so there are only 2 commits in total. Then I'm good to merge it into master.
We'll do a new release then.

juliusv commented:
> For the future we may want to look at https://github.com/golang/dep for dependency management/vendoring. Do you know this project?

dep was built by a group of experienced Go people and once slated to become the official vendoring solution for Go. But then some people from the original Go team found that they were not happy with some aspects of dep and created yet another tool, vgo (https://github.com/golang/vgo, https://golang.org/issue/24301), that the Go team now decided should become the standard Go vendoring tool. There was a big community uprising against that, as there were people unhappy with the way vgo resolves versions of dependencies, and it all turned into a religious war... so the final outcome isn't really clear yet, except that vgo probably has a higher probability of actually turning into the "official" vendoring tool that will be built into Go itself. Personally I haven't used vgo yet, but have used dep and think it's a very solid solution. When using dep you should be aware though that it takes much longer to run than say "govendor", since it does more sophisticated dependency resolution and then downloads the necessary package versions on every vendoring update, rather than just copying what you have in your GOPATH somewhere.

In summary: it shouldn't matter too much, you can switch easily between vendoring tools, the current govendor works fine enough too, but dep might be a more robust solution, vgo might become the standard though, but it's early :)

juliusv commented:
@chaudum Squashed the commits as requested (now there's a vendoring commit and one with all the other changes).

juliusv commented:
I added tests for the configuration loading (which actually found some issues around YAML parsing and defaults). I squashed it all again immediately so it'd be ready to merge, I hope that's ok.

chaudum commented:
@juliusv Thanks a lot!
